crates/host/src/wasmbus/mod.rs

```diff
@@ -2328,36 +2356,68 @@ impl Host {
-                        json!({
-                            "public_key": public_key,
-                            "link_name": link_name,
-                            "contract_id": contract_id.unwrap_or("n/a"),
-                        })
-                    })
+                    let jwt::CapabilityProvider {
+                        capid: contract_id,
+                        name,
+                        rev: revision,
+                        ..
+                    } = claims.metadata.as_ref()?;
+                    Some(instances.iter().map(
+                        move |(link_name, ProviderInstance { annotations, .. })| {
+                            let annotations = Some(annotations.clone().into_iter().collect());
+                            let revision = revision.unwrap_or_default();
+                            ProviderDescription {
+                                id: public_key.into(),
+                                image_ref: Some(image_ref.clone()),
+                                contract_id: contract_id.clone(),
+                                link_name: link_name.into(),
+                                name: name.clone(),
+                                annotations,
+                                revision,
+                            }
+                        },
+                    ))
                 },
             )
+            .flatten()
             .collect();
         let uptime = self.start_at.elapsed();
-        json!({
-            "actors": actors,
-            "friendly_name": self.friendly_name,
-            "labels": *self.labels.read().await,
-            "providers": providers,
-            "uptime_human": human_friendly_uptime(uptime),
-            "uptime_seconds": uptime.as_secs(),
-            "version": env!("CARGO_PKG_VERSION"),
-        })
+        HostInventory {
+            actors,
+            providers,
+            friendly_name: self.friendly_name.clone(),
+            labels: self.labels.read().await.clone(),
+            uptime_human: human_friendly_uptime(uptime),
+            uptime_seconds: uptime.as_secs(),
+            version: env!("CARGO_PKG_VERSION").to_string(),
+            host_id: self.host_key.public_key(),
+            issuer: self.cluster_key.public_key(),
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn heartbeat(&self) -> serde_json::Value {
+        trace!("generating heartbeat");
+        let inventory = self.inventory().await;
+        match serde_json::to_value(inventory) {
+            Ok(value) => value,
+            // SAFETY: Failing to serialize the inventory shouldn't be possible as we control the instantiation of the
+            // struct For safety we'll fall back on the default but that shouldn't ever occur.
+            Err(e) => {
+                error!(%e, "failed to serialize inventory for heartbeat");
+                serde_json::Value::default()
```

Comment:
If we can't serialize the inventory, I suspect it's better to not publish anything, since this will (probably?) break any client trying to parse the heartbeat payload

Reply:
Good call